### PR TITLE
[perf] [1/n] enhance 'request-pending' logic in consensus_peer

### PR DESCRIPTION
**Pull request by bhatvinay**
Summary:
Enhancements to "request_pending" tracker of the Peer:
For every peer, the leader tracks if a request is pending i.e a request is in
flight for this particular peer. There can only be one outstanding request for
a peer. This PR changes the semantics of 'request_pending' to mean 'request
is being built and/or request is in flight' (instead of 'request is in flight'). Building
a request could be expensive for a lagging peer as it need to do disk IO. This
could block threads which are holding onto RaftConsensus::lock_ and
PeerManager::lock_, thus blocking out the commit process completely.
This PR makes it so that when a request is being built (i.e
Peer::SendNextRequest(...) calls into PeerMessageQueue::RequestForPeer()), we
return early when attempting to build another request for the same peer.
The rationale is that if there are multiple calls into
Peer::SendNextRequest() (from different callsites), then there is no
need to build duplicate requests for this particular peer. Any
additional operations can always be shipped in a subsequent batch. To
make this check lock-free, "request_pending_" is made an atomic.
This ensures that RaftConsensus::lock_ and PeerManager::lock_ are not locked
for an extended duration

Test Plan: fbcode tests and perf tests

Reviewers: arahut

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by bhatvinay** [comment]
Benchmarks done with prod like workload showed that this helps with reducing commit stalls when the leader has to catchup a lagging peer. Commit stalls reduce by ~70% (i.e there are 70% lower commit stalls)

**Comment by bhatvinay** [comment]
This was split from PR131 which also had the logic to start splitting underlying data structures - we will revisit the splitting of the data structures later

File: src/kudu/consensus/raft_consensus.cc [hunk 1]
```diff
@@ -3037,7 +3038,9 @@ Status RaftConsensus::RefreshConsensusQueueAndPeersUnlocked() {
   queue_->SetLeaderMode(pending_->GetCommittedIndex(),
                         CurrentTermUnlocked(),
                         active_config);
-  RETURN_NOT_OK(peer_manager_->UpdateRaftConfig(active_config));
+  RETURN_NOT_OK(peer_manager_->UpdateRaftConfig(
+        active_config,
+        queue_->GetQuorumMode()));
```
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc]: this seems to be a remnant due to the PR splitting, where in PeerManager you needed quorum mode

**Review by anirbanr-fb — COMMENTED** [review]
The changes look good!

The patch comment was a bit confusing to me.

"" This PR changes the semantics of 'request_pending' to mean 'request
is being built' (instead of 'request is in flight').""

I think request_pending_ now means that the request is being built or it is in flight. both

**Review by anirbanr-fb — APPROVED** [review]

File: src/kudu/consensus/raft_consensus.cc (hunk 1 above)
Comment (bhatvinay) [reply src/kudu/consensus/raft_consensus.cc]: Yes. Fixed it and running another round of tests and benchmarks before merging

**Comment by bhatvinay** [comment]
> The patch comment was a bit confusing to me.
> 
> "" This PR changes the semantics of 'request_pending' to mean 'request is being built' (instead of 'request is in flight').""
> 
> I think request_pending_ now means that the request is being built or it is in flight. both

Updated the comments to reflect the fact that request_pending means 'request is being built OR request is in flight'

**Pull request merged**